File: shortenUrl/src/main/java/com/oscar/shortenurl/controller/ShortenController.java
package com.oscar.shortenurl.controller;

import com.oscar.shortenurl.service.ShortenService;
import lombok.AllArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@AllArgsConstructor
public class ShortenController {

    private ShortenService shortenService;

    @GetMapping("/hello")
    public String hello() {
        return "hello";
    }

    @PostMapping("/shorten")
    public ResponseEntity<String> postShorten(@RequestBody Map<String, String> originalUrl) {
        try {
            String url = shortenService.saveShortenUrl(originalUrl.get("url"));
            return ResponseEntity
                    .ok()
                    .body(url);
        } catch (Exception e) {
            String error = "Unknown Error\n" + e;
            return ResponseEntity
                    .status(HttpStatus.SERVICE_UNAVAILABLE)
                    .body(error); // 서버 에러를 반환하는 좋은 방법?
        }
    }

    @GetMapping("/redirect")
    public ResponseEntity<String> getRedirect(@RequestBody Map<String, String> shortenUrl) throws Exception {
        try {
            String url = shortenService.getOriginalUrl(shortenUrl.get("url"), true);
            return ResponseEntity
                    .status(HttpStatus.PERMANENT_REDIRECT)
                    .header("Location", url)
                    .build();
        } catch (Exception e) {
            String error = "Unknown Error\n" + e;
            return ResponseEntity
                    .status(HttpStatus.SERVICE_UNAVAILABLE)
                    .body(error); // 서버 에러를 반환하는 좋은 방법?
        }
    }

    @GetMapping("/redirect/{path}")
    public ResponseEntity<String> getRedirect(@PathVariable String path) {
        try {
            String url = shortenService.getOriginalUrl(path, false);
            return ResponseEntity
                    .status(HttpStatus.TEMPORARY_REDIRECT)
                    .header("Location", url)
                    .build();
        } catch (Exception e) {
            String error = "Unknown Error\n" + e;
            return ResponseEntity
                    .status(HttpStatus.SERVICE_UNAVAILABLE)
                    .body(error); // 서버 에러를 반환하는 좋은 방법?
        }
    }

    @GetMapping("/redirect/{path}/count")
    public ResponseEntity<String> getRedirectCount(@PathVariable String path) {
        try {
            long count = shortenService.getRedirectCount(path);
            return ResponseEntity
                    .ok()
                    .body(path + " redirect count: " + count);
        } catch (Exception e) {
            String error = "Unknown Error\n" + e;
            return ResponseEntity
                    .status(HttpStatus.SERVICE_UNAVAILABLE)
                    .body(error); // 서버 에러를 반환하는 좋은 방법?
        }
    }
}

File: kotlin/shortenUrl/src/main/kotlin/oscar/kotlin/shortenUrl/controller/ShortenController.kt
package oscar.kotlin.shortenUrl.controller

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import oscar.kotlin.shortenUrl.mapper.ShortenMapper
import oscar.kotlin.shortenUrl.model.*
import oscar.kotlin.shortenUrl.service.ShortenService

@Controller
class ShortenController (
    private val shortenService: ShortenService,
    private val shortenMapper: ShortenMapper
    ) {

    @PostMapping("/shorten")
    fun postShorten(@RequestBody originalUrl: ShortenRequestModel) : ResponseEntity<ShortenResponseModel> {
        val urlDto = shortenService.saveShortenUrl(shortenMapper.toShortenRequestDto(originalUrl))
        val toShortenResponseModel = shortenMapper.toShortenResponseModel(urlDto)
        return ResponseEntity.ok(toShortenResponseModel)
    }

    @GetMapping("/redirect/{path}")
    fun getRedirect(@PathVariable path: RedirectRequestModel): ResponseEntity<Any> {
        val url: String = shortenService.getOriginalUrl(path.shortenUrl)
        return ResponseEntity
            .status(HttpStatus.PERMANENT_REDIRECT)
            .header("Location", url)
            .build()
    }

    @GetMapping("/redirect/{path}/count")
    fun getRedirectCount(@PathVariable path: RedirectCountRequestModel) : ResponseEntity<RedirectCountResponseModel> {
        val redirectCountDto = shortenService.getRedirectCount(path.shortenUrl)
        val toRedirectCountResponseModel = shortenMapper.toRedirectCountResponseModel(redirectCountDto)
        return ResponseEntity.ok(toRedirectCountResponseModel)
    }
}